BCStudentSoftwareDevTeam/celts
# Minor integration

### New pages for CCE Minor: 
CCE Minor Profile page: `/profile/<username>/cceMinor`
Request other engagement for CCE minor: `/cceMinor/<username>/requestOtherCommunityEngagement`
CCE Minor admin page: `/admin/cceMinor`

### Model changes: 
`certification`
- Changed `CESC` to `CCE`

`communityEngagementRequest`
- Added `company` field
- Allow `filename` to be null

Added `individualRequirement` table
- The table matches a program or course to its respective requirement in the `certificationRequirement` table

Added `otherExperience` table
- The table will contain all of the information for an other experience outside of normal programs/courses that will be credited to a Sustained Community Engagement

### To Test:
- Checkout branch `minorIntegration` and run `database/reset_database.sh` followed by `tests/run_tests.sh` 
- As long as all the tests pass run `flask run` next.
- Navigate to a student's profile, `/profile/<username>`. If the student has no credit towards the CCE minor then there should be a notification box in the top right of their profile asking if they are interested in pursuing the CCE minor and if they do have progress towards the minor then that notification box should not be there. 

- Next navigate to the CCE Minor profile page, `/profile/<username>/cceMinor`
- On the Minor profile page verify all the functionality works. 
1. If the student has a Course Participation record or an Event Participation record the term in which that record is in should have its accordion. Inside this accordion should be the name of the Course or Program that the event was under. You should be able to click on the row to see more information about the Course/Events the student has participated in. 
2. You should be able to add Courses or Programs as credit for Sustained Community engagement by clicking the checkbox. 
3. You should be able to add 4 Courses or Programs for credit and if you try to add a 5th you should be told the student already has 4 credits and can not have others added.  
4. After a student has credit for 4 different things credited towards Sustained Community engagement you should be able to refresh the page and have a checkmark appear which should indicate they have completed all Sustained Community engagements. 

- Next, test the Summer Engagement functionality. 
- The cases to check are: 
1. If a student does not have any Summer Engagement record then the text box should be empty with the placeholder and if you Save or Clear with the empty text box nothing should break 
2. Enter a Summer Engagement for a student and reload the page. The record and term you put in should both still be what you entered and there should be a check next to the Summer Engagement header indicating they have completed the Summer Engagement.
3. Clear the Summer Engagement and verify the record was cleared successfully. 

-Next, click the `Request Other Experience` button. You should be taken to the following route: `cceMinor/<username>/requestOtherCommunityEngagement`. 
- Fill in all the information that is requested on the page and verify everything is saved in the database correctly. 
- Make sure when you upload a file it is saved to the correct place, `app/static/files` 

-Finally, navigate to the Admin Minor management page, `/admin/cceMinor`
- Verify all the students who have progress in the minor are in the `Minor Progress` table, all of the counts are correct, and all of the links to the respective students profiles are correct. 
- Verify all the students in the `Interested Student` table are interested, and the names link to the correct profiles.
